File: client/src/types/resume.ts
// src/types/resume.ts
// -----------------------------------------------------------------------------
// Adds yearsByCategory (per-domain breakdown) and primaryYears (recommended)
// while keeping yearsOfExperience for backward-compat.
// -----------------------------------------------------------------------------

export type SkillItem = {
  name: string;
  source: string;
  weight: number;
  category: string | null;
};

export type ApiResumeSummary = {
  id: string;
  name: string | null;
  profession: string | null;
  years_of_experience: number | null;
  resume_url: string;
  // NEW:
  years_by_category?: Record<string, number>; // optional on summary
};

export type ApiResumeListResponse = {
  items: ApiResumeSummary[];
  total: number;
};

export type ApiResumeDetail = {
  id: string;
  name: string | null;
  profession: string | null;
  years_of_experience: number | null;
  resume_url: string;
  status: string;
  file_name: string | null;
  mime_type: string | null;
  file_size: number | null;
  summary: string | null;
  contacts: Array<{
    type: string;
    label: string | null;
    value: string;
  }>;
  skills: SkillItem[];
  experience: Array<{
    title: string | null;
    company: string | null;
    location: string | null;
    start_date: string | null;
    end_date: string | null;
    bullets?: string[];
    tech?: string[];
    duration_years: number | null;
  }>;
  education: Array<{
    degree: string | null;
    field: string | null;
    institution: string | null;
    start_date: string | null;
    end_date: string | null;
  }>;
  languages: string[];
  created_at: string; // ISO
  updated_at: string; // ISO
  // NEW:
  years_by_category?: Record<string, number>;
  primary_years?: number | null;
};

export type ResumeSummary = {
  id: string;
  name: string | null;
  profession: string | null;
  yearsOfExperience: number | null;
  resumeUrl: string;
  // NEW:
  yearsByCategory: Record<string, number>;
};

export type ResumeListResponse = {
  items: ResumeSummary[];
  total: number;
};

export type ResumeDetail = {
  id: string;
  name: string | null;
  profession: string | null;
  yearsOfExperience: number | null;
  resumeUrl: string;
  status: string;
  fileName: string | null;
  mimeType: string | null;
  fileSize: number | null;
  summary: string | null;
  contacts: Array<{ type: string; label: string | null; value: string }>;
  skills: SkillItem[];
  experience: Array<{
    title: string | null;
    company: string | null;
    location: string | null;
    startDate: string | null;
    endDate: string | null;
    bullets: string[];
    tech: string[];
    durationYears: number | null;
  }>;
  education: Array<{
    degree: string | null;
    field: string | null;
    institution: string | null;
    startDate: string | null;
    endDate: string | null;
  }>;
  languages: string[];
  createdAt: string; // ISO
  updatedAt: string; // ISO
  yearsByCategory: Record<string, number>;
  primaryYears: number | null;
};

File: backend/app/schemas/resume.py
# app/schemas/resume.py  (the Pydantic models file you provided)
# -----------------------------------------------------------------------------
# CHANGELOG (English-only comments as requested)
# - Add optional years_by_category: dict[str, float] to both ResumeSummary and ResumeDetail.
# - Add optional primary_years: float to ResumeDetail (recommended primary years, e.g., "tech").
# - Keep years_of_experience for backward compatibility (legacy total).
# - No breaking changes: all new fields are optional and default to None/{}.
# -----------------------------------------------------------------------------
from __future__ import annotations

from datetime import datetime
from typing import Optional, Dict
from uuid import UUID

from pydantic import BaseModel, Field


class SkillItem(BaseModel):
    """Skill with source tracking for weighted matching.

    Simplified weighting model (binary groups):
    - Experience skill (appears in work experience context): weight = 1.0, source typically "work_experience".
    - General skill (projects/education/deterministic/list/etc.): weight = 0.6, source varies ("projects", "education", "skills_list", "deterministic", etc.).

    Consumers should branch only by weight (==1.0 => experience, else general). Legacy sources retained for backward compatibility.
    """
    name: str
    source: str = "skills_list"  # legacy values preserved; grouping now by weight only
    weight: float = Field(default=0.6, ge=0.0, le=1.0)  # 1.0=experience, 0.6=general
    category: Optional[str] = None  # Optional classification (language/framework/tool/etc.)


class ResumeSummary(BaseModel):
    id: UUID
    name: Optional[str] = None
    profession: Optional[str] = None
    years_of_experience: Optional[float] = Field(None, ge=0)
    resume_url: str
    # New: per-category breakdown (optional)
    years_by_category: Dict[str, float] = Field(default_factory=dict)


class ResumeListOut(BaseModel):
    items: list[ResumeSummary]
    total: int


class ResumeContactItem(BaseModel):
    type: str
    label: Optional[str] = None
    value: str


class ResumeExperienceItem(BaseModel):
    title: Optional[str] = None
    company: Optional[str] = None
    location: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    bullets: list[str] = Field(default_factory=list)
    tech: list[str] = Field(default_factory=list)
    duration_years: Optional[float] = Field(None, ge=0)


class ResumeEducationItem(BaseModel):
    degree: Optional[str] = None
    field: Optional[str] = None
    institution: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None


class ResumeDetail(BaseModel):
    id: UUID
    name: Optional[str] = None
    profession: Optional[str] = None
    years_of_experience: Optional[float] = Field(None, ge=0)
    resume_url: str
    status: str
    file_name: Optional[str] = None
    mime_type: Optional[str] = None
    file_size: Optional[int] = None
    summary: Optional[str] = None
    contacts: list[ResumeContactItem] = Field(default_factory=list)
    skills: list[SkillItem] = Field(default_factory=list)
    experience: list[ResumeExperienceItem] = Field(default_factory=list)
    education: list[ResumeEducationItem] = Field(default_factory=list)
    languages: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    years_by_category: Dict[str, float] = Field(default_factory=dict)
    primary_years: Optional[float] = Field(None, ge=0)
